### ISSUE TITLE
Swish activation function

### ISSUE THREAD

**Issue by gigasquid** [thinktopic/cortex#242]
Saw this paper the other day and it looks like it could be a useful alternative to relu

https://arxiv.org/abs/1710.05941

**Comment by harold** [comment]
Funny, we were just talking about this today in the office.

It is very reul-y:
http://www.wolframalpha.com/input/?i=x+*+sigmoid+(x)

If you grep the code for `relu` you'd find most of the places you'd need to touch.

CPU side would likely be very straight forward. On CUDA we use cuDNN activation functions. Not sure if they're going to have swish, if not, implementing it would involve writing a tensor operation (Chris will correct me if I'm wrong).

Could be fun to implement, and would definitely be fun to compare to our other activation functions.

**Comment by gigasquid** [comment]
I've had some fun working on this. I had a question on how to best implement the new activations.
It was pretty straightforward doing the cpu side of things, but when I got to the gpu, I had a question of how to best implement it. The way I see it I could:

1) Write high level tensor operations to express the swish forward and backward in terms of other existing unary operations. This implementation would work the same across the cpu and gpu. Example of the approach is here https://github.com/gigasquid/cortex/blob/swish/src/cortex/compute/nn/layers.clj#L91

2) Write a new unary operation `:swish` that is implemented differently in the cpu and gpu. In  the gpu side of things it would have an implementation in terms of the existing unary operations.

3) Like #2 - But on the gpu side that would be ultimately implemented in the low level operations file https://github.com/thinktopic/cortex/blob/master/resources/operations.h

Let me know your thoughts.

**Comment by gigasquid** [comment]
I realized it would most likely be easier if you could see all the code so I opened a PR for it. Feedback is definitely welcome and I'm open to alternative implementations.

**Comment by harold** [comment]
@cnuernber --- relevant to your interests.

**Comment by cnuernber** [comment]
I landed the PR.

I personally wouldn't do a custom tensor up unless there was a reason like CUDNN had a fast-path that combined those operations in some special way.  Or if we started hitting an issue where those were the most expensive operations in some network.  

That being said, a custom operator would be good if you just wanted to learn how things work.  But I don't see it as being necessary aside from the learning potential.
